microprofile.jdt/org.eclipse.lsp4mp.jdt.core/src/main/java/org/eclipse/lsp4mp/jdt/internal/faulttolerance/java/MicroProfileFaultToleranceASTValidator.java

```diff
@@ -204,6 +230,29 @@ private void validateAsynchronousAnnotation(MethodDeclaration node, MarkerAnnota
 		}
 	}
 
+	/**
+	 * Checks if the given method declaration has a retry annotation, and if so,
+	 * provides diagnostics for the delay and maxDuration value(s)
+	 *
+	 * @param node       The method declaration to validate
+	 * @param annotation The @Retry annotation
+	 * @throws JavaModelException
+	 */
+	private void validateRetryAnnotation(MethodDeclaration node, NormalAnnotation annotation)
+			throws JavaModelException {
+		Expression delayExpr = getAnnotationMemberValueExpression(annotation, DELAY_RETRY_ANNOTATION_MEMBER);
+		Expression maxDurationExpr = getAnnotationMemberValueExpression(annotation,
+				MAX_DURATION_RETRY_ANNOTATION_MEMBER);
+		if (delayExpr != null && maxDurationExpr != null) {
+			int delayValue = Integer.parseInt(delayExpr.toString());

```

Comment:
I don't think the way to do it is by using toString() directly. An Expression can be a StringLiteral, or a NumberLiteral, etc. I think you need to check that it's the one you expect, and then get the value from that.

Reply:
Without toString(), parseInt won't work since Java doesn't seem to recognize an Expression as a StringLiteral. Should I use a cast instead?

Reply:
Actually, I think the value would be a NumberLiteral, so you would have to check for that. Either way, if you insert a breakpoint here and check the type of Expression, it should tell you. Then you just have to find what method to use in order to get at the value.

Reply:
It seems when it's a negative number, it's classified as a PrefixExpression, and when it's a positive number, it's considered a NumberLiteral. However in the Java code, it is always assigned as an Expression, which will throw a Syntax error when using any operator that accepts a number.

Reply:
But we don't need to support negative numbers for this particular validation. In fact if it were a negative number, you could safely fail the validation. Did you check whether Expression has any methods that might be of use in getting to something you can use ?

Reply:
Expression has `resolveConstantExpressionValue()`, which should work out for what we want to implement.

For supporting negative values, if we have 
```java
@Retry(delay = -1000, maxDuration = -5000)
```
we should throw the negative diagnostic, but should we also throw the `DELAY_EXCEEDS_MAX_DURATION` error (which is supported atm)?

Reply:
We shouldn't throw the delay exceeding max duration is only really defined and meaningful when we're deal with positive numbers. If any of those values are negative, then the first thing to address is the negative value.